auth/end-user/web/main.py

```diff
@@ -0,0 +1,119 @@
+# Copyright 2017 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""An example web application that obtains authorization and credentials from
+an end user.
+
+This sample is used on
+https://developers.google.com/identity/protocols/OAuth2WebServer. Please
+refer to that page for instructions on using this sample.
+
+Notably, you'll need to obtain a OAuth2.0 client secrets file and set the
+``GOOGLE_CLIENT_SECRETS`` environment variable to point to that file.
+"""
+
+import os
+
+import flask
+import google.oauth2.credentials
+import google_auth_oauthlib.flow
+import googleapiclient.discovery
+
+# The path to the client-secrets.json file obtained from the Google API
+# Console. You must set this before running this application.
+CLIENT_SECRETS_FILENAME = os.environ['GOOGLE_CLIENT_SECRETS']
+# The OAuth 2.0 scopes that this application will ask the user for. In this
+# case the application will ask for basic profile information.
+SCOPES = ['email', 'profile']
+
+app = flask.Flask(__name__)
+# TODO: A secret key is included in the sample so that it works but if you
+# use this code in your application please replace this with a truly secret
+# key. See http://flask.pocoo.org/docs/0.12/quickstart/#sessions.
+app.secret_key = 'TODO: replace with a secret value'
+
+
+@app.route('/')
+def index():
+    if 'credentials' not in flask.session:
+        return flask.redirect('authorize')
+
+    # Load the credentials from the session.
+    credentials = google.oauth2.credentials.Credentials(
+        **flask.session['credentials'])
```

Comment:
Is there a way to check credential validity, either here or after issuing the request if it failed?

Reply:
`credentials.valid` but the apiclient will refresh if needed and raise a useful error if it's unrefreshable/otherwise invalid.